[VPlan] Introduce recipes for VP loads and stores.

fhahn opened the pull request:
Introduce new subclasses of VPWidenMemoryRecipe for VP
(vector-predicated) loads and stores to address multiple TODOs from
https://github.com/llvm/llvm-project/pull/76172

Note that the introduction of the new recipes also improves code-gen for
VP gather/scatters by removing the redundant header mask. With the new
approach, it is not sufficient to look at users of the widened canonical
IV to find all uses of the header mask.

In some cases, a widened IV is used instead of separately widening the
canonical IV. To handle those cases, iterate over all recipes in the
vector loop region to make sure all widened memory recipes are
processed.

Depends on https://github.com/llvm/llvm-project/pull/87411.

fhahn commented:
Note that this still contains the commit for https://github.com/llvm/llvm-project/pull/87411 as that PR is based on a branch not in `llvm-project`. Will update once https://github.com/llvm/llvm-project/pull/87411 is merged.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9535,6 +9471,51 @@ void VPWidenLoadRecipe::execute(VPTransformState &State) {
+  // Handle loads.
+  assert(LI && "Must have a load instruction");
+  State.setDebugLocFrom(getDebugLoc());
+  for (unsigned Part = 0; Part < State.UF; ++Part) {
```
No need for the loop here, since expected that State.UF == 1

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9603,6 +9567,48 @@ void VPWidenStoreRecipe::execute(VPTransformState &State) {
+  auto &Builder = State.Builder;
+  State.setDebugLocFrom(getDebugLoc());
+
+  for (unsigned Part = 0; Part < State.UF; ++Part) {
```
Same here

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2392,7 +2391,51 @@ struct VPWidenLoadRecipe final : public VPWidenMemoryRecipe, public VPValue {
+  }
+
+  VPRecipeBase *clone() override {
+    return new VPWidenVPLoadRecipe(cast<LoadInst>(Ingredient), getAddr(),
+                                   getEVL(), getMask(), isConsecutive(),
+                                   getDebugLoc());
```
Do we need clone implementation?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1347,6 +1337,48 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
   // TODO: support unroll factor > 1.
   Plan.setUF(1);
+
+  VPValue *BTC = Plan.getOrCreateBackedgeTakenCount();
+  auto IsHeaderMask = [BTC](VPValue *V) {
```
```suggestion
  auto IsHeaderMask = [BTC = Plan.getOrCreateBackedgeTakenCount()](VPValue *V) {
```

ayalz reviewed (commented):
Thanks for taking care of this clean-up!

nikolaypanchenko commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2372,15 +2391,144 @@ class VPWidenMemoryInstructionRecipe : public VPRecipeBase {
+/// A recipe for widening load operations with vector-predication intrinsics,
+/// using the address to load from, the explicit vector length and an optional
+/// mask.
+struct VPWidenVPLoadRecipe final : public VPWidenMemoryRecipe, public VPValue {
```
My understanding that you want to have dedicated recipes that emit vp-intrinsics. Since they're derived from VPRecipe, is it expected they will be fully supported by VPlanTransforms ?

fhahn committed:
[VPlan] Introduce recipes for VP loads and stores.

Introduce new subclasses of VPWidenMemoryRecipe for VP
(vector-predicated) loads and stores to address multiple TODOs from
https://github.com/llvm/llvm-project/pull/76172

Note that the introduction of the new recipes also improves code-gen for
VP gather/scatters by removing the redundant header mask. With the new
approach, it is not sufficient to look at users of the widened canonical
IV to find all uses of the header mask.

In some cases, a widened IV is used instead of separately widening the
canonical IV. To handle those cases, iterate over all recipes in the
vector loop region to make sure all widened memory recipes are
processed.

Depends on https://github.com/llvm/llvm-project/pull/87411.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1348,6 +1337,48 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+    return CompareToReplace &&
+           CompareToReplace->getOpcode() == Instruction::ICmp &&
+           CompareToReplace->getPredicate() == CmpInst::ICMP_ULE &&
+           CompareToReplace->getOperand(1) == BTC;
```
> Must every ICMP_ULE of BTC be the header mask, regardless of its other operand?

At the moment yes, but I reworked this to walk users of header-masks.

The current placement of `addExplicitVectorLength` unfortunately means that `VPWidenCanonicalIVRecipe` may have been removed, if there's a matching `VPWidenIntOrFpInductionRecipe` (see TODO from the initial EVL PR; the current placement is to avoid introducing new uses of the canonical IV after introducing EVL)

So we also need to walk the users of `VPWidenIntOrFpInductionRecipe` that are canonical.

> Tail folding (an integral part of initial stripmining step according to roadmap) should arguably introduce a single abstract HeaderMask VPValue, retrievable from the loop region (as with its canonical IV), to be later materialized/legalized/lowered into concrete bump/widening/compare/active-lane-mask/no-mask-with-EVL recipes placed inside blocks?

Sounds good to me, pushing towards more gradual lowering (as follow-up)

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1348,6 +1337,48 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+      if (!MemR)
+        continue;
+      VPValue *OrigMask = MemR->getMask();
+      if (!OrigMask)
```
Should not be null at this point, updated, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1348,6 +1337,48 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+        continue;
+      assert(!MemR->isReverse() &&
+             "Reversed memory operations not supported yet.");
+      VPValue *Mask = IsHeaderMask(OrigMask) ? nullptr : OrigMask;
```
Reordered, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1348,6 +1337,48 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+             "Reversed memory operations not supported yet.");
+      VPValue *Mask = IsHeaderMask(OrigMask) ? nullptr : OrigMask;
+      if (auto *L = dyn_cast<VPWidenLoadRecipe>(&R)) {
+        auto *N = new VPWidenVPLoadRecipe(cast<LoadInst>(L->getIngredient()),
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1348,6 +1337,48 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+        L->replaceAllUsesWith(N);
+        L->eraseFromParent();
+      } else if (auto *S = dyn_cast<VPWidenStoreRecipe>(&R)) {
+        auto *N = new VPWidenVPStoreRecipe(
```
Done, thanks!

fhahn reviewed (commented):
Rebased after parent commits landed & comments addressed. Apologies for the force-push

ayalz commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-gather-scatter.ll:
```diff
@@ -47,17 +44,16 @@ define void @gather_scatter(ptr noalias %in, ptr noalias %out, ptr noalias %inde
-; IF-EVL-NEXT:    [[TMP19:%.*]] = icmp ule <vscale x 2 x i64> [[VEC_IND]], [[BROADCAST_SPLAT]]
 ; IF-EVL-NEXT:    [[TMP20:%.*]] = getelementptr inbounds i32, ptr [[INDEX:%.*]], <vscale x 2 x i64> [[VEC_IND]]
-; IF-EVL-NEXT:    [[WIDE_MASKED_GATHER:%.*]] = call <vscale x 2 x i64> @llvm.vp.gather.nxv2i64.nxv2p0(<vscale x 2 x ptr> align 8 [[TMP20]], <vscale x 2 x i1> [[TMP19]], i32 [[TMP18]])
+; IF-EVL-NEXT:    [[WIDE_MASKED_GATHER:%.*]] = call <vscale x 2 x i64> @llvm.vp.gather.nxv2i64.nxv2p0(<vscale x 2 x ptr> align 8 [[TMP20]], <vscale x 2 x i1> shufflevector (<vscale x 2 x i1> insertelement (<vscale x 2 x i1> poison, i1 true, i64 0), <vscale x 2 x i1> poison, <vscale x 2 x i32> zeroinitializer), i32 [[TMP18]])
```
nit: would be clearer if this `shufflevector ( ... zeroinitializer)` operand representing a full-mask was outlined; as some "oneinitializer".

ayalz replied:
ok

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1336,6 +1332,30 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
   NextEVLIV->insertBefore(CanonicalIVIncrement);
   EVLPhi->addOperand(NextEVLIV);
 
+  forAllHeaderPredicates(Plan, [VPEVL](VPInstruction &Mask) {
```
An alternative approach would be to scan the recipes inside the loop, replacing every widen memory recipe with its vp counterpart, where header masks are replaced with null. Before doing so, collect all header masks into a set, possibly via forAllHeaderPredicates(), which could be simplified into collectAllHeaderPredicates(), until a designated recipe can be use directly.

Where is `collectUsersRecursively()` defined?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9425,6 +9362,44 @@ void VPWidenLoadRecipe::execute(VPTransformState &State) {
   }
 }
 
+void VPWidenVPLoadRecipe::execute(VPTransformState &State) {
```
A general naming comment: `VPWidenVP...` sounds a bit confusing. The original non-EVL/non-VP masked vector loads and stores are also already "vector predicated", and the VP prefix is already prevalently used to denote VPlan. The distinction lies with the additional EVL operand, currently employed to relieve the mask of conveying the tail. How about `VPWidenEVLLoadRecipe` and `VPWidenEVLStoreRecipe`?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-split-vpmemory

fhahn committed:
!fixup address latest comments, thanks!

fhahn reviewed (commented):
Thanks for taking a look, latest comments should be addressed.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -91,6 +94,7 @@ bool VPRecipeBase::mayReadFromMemory() const {
   case VPPredInstPHISC:
   case VPScalarIVStepsSC:
   case VPWidenStoreSC:
+  case VPWidenEVLStoreSC:
```
lex order.

fhahn replied:
updated with new naming

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -157,6 +161,8 @@ bool VPRecipeBase::mayHaveSideEffects() const {
   case VPWidenLoadSC:
   case VPWidenStoreSC:
+  case VPWidenEVLLoadSC:
+  case VPWidenEVLStoreSC:
```
lex order.

fhahn replied:
updated with new naming

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1203,43 +1203,54 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
-      continue;
+  for (auto *Wide : WideCanonicalIVs) {
+    for (VPUser *U : SmallVector<VPUser *>(Wide->users())) {
+      // TODO: Introduce explicit recipe for header-mask instead of searching
+      // for the header-mask pattern manually.
```
This TODO can be placed at the definition of collectAllHeaderMasks(), which would be removed as a whole.

fhahn replied:
Moved, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1336,6 +1340,31 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+        auto *N = new VPWidenEVLLoadRecipe(L, VPEVL, NewMask);
+        N->insertBefore(L);
+        L->replaceAllUsesWith(N);
+        L->eraseFromParent();
```
```suggestion
```
nit: could the following `recursivelyDeleteDeadRecipes()` take care of erasing Load recipes from their parents? The former is needed due to no subsequent VPlan dce.
Collecting dead Stores is more challenging.

fhahn replied:
At the moment `recursivelyDeleteDeadRecipes` works upwards, recursing from uses to def, so `recursivelyDeleteDeadRecipes(HeaderMask)` won't clean up the load.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -358,6 +358,8 @@ class VPDef {
     VPWidenLoadSC,
     VPWidenStoreSC,
+    VPWidenEVLLoadSC,
+    VPWidenEVLStoreSC,
```
lex order.

fhahn replied:
reordered with new naming, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-split-vpmemory

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2390,13 +2387,48 @@ struct VPWidenLoadRecipe final : public VPWidenMemoryRecipe, public VPValue {
+struct VPWidenEVLLoadRecipe final : public VPWidenMemoryRecipe, public VPValue {
+  VPWidenEVLLoadRecipe(VPWidenLoadRecipe *L, VPValue *EVL, VPValue *Mask)
+      : VPWidenMemoryRecipe(
+            VPDef::VPWidenEVLLoadSC, *cast<LoadInst>(&L->getIngredient()),
```
Ah no, removed, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2436,6 +2468,51 @@ struct VPWidenStoreRecipe final : public VPWidenMemoryRecipe {
+struct VPWidenEVLStoreRecipe final : public VPWidenMemoryRecipe {
+  VPWidenEVLStoreRecipe(VPWidenStoreRecipe *S, VPValue *EVL, VPValue *Mask)
+      : VPWidenMemoryRecipe(VPDef::VPWidenEVLStoreSC,
+                            *cast<StoreInst>(&S->getIngredient()),
```
Not needed, removed, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -48,6 +48,7 @@ bool VPRecipeBase::mayWriteToMemory() const {
   case VPInterleaveSC:
     return cast<VPInterleaveRecipe>(this)->getNumStoreOperands() > 0;
   case VPWidenStoreSC:
+  case VPWidenEVLStoreSC:
```
Renamed including the recipe names (for consistency), thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -65,6 +66,7 @@ bool VPRecipeBase::mayWriteToMemory() const {
   case VPWidenIntOrFpInductionSC:
   case VPWidenLoadSC:
   case VPWidenPHISC:
+  case VPWidenEVLLoadSC:
```
Updated with new naming.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -82,6 +84,7 @@ bool VPRecipeBase::mayWriteToMemory() const {
 bool VPRecipeBase::mayReadFromMemory() const {
   switch (getVPDefID()) {
   case VPWidenLoadSC:
+  case VPWidenEVLLoadSC:
```
Updated with new naming

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1203,43 +1203,54 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
-static void
-replaceHeaderPredicateWith(VPlan &Plan, VPValue &Idiom,
-                           function_ref<bool(VPUser &, unsigned)> Cond = {}) {
+/// Collet all VPValues representing a header mask through the (ICMP_ULE,
```
Fixed, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1203,43 +1203,54 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
+  if (FoundWidenCanonicalIVUser != Plan.getCanonicalIV()->users().end()) {
+    auto *WideCanonicalIV =
+        cast<VPWidenCanonicalIVRecipe>(*FoundWidenCanonicalIVUser);
+    assert(all_of(Plan.getCanonicalIV()->users(),
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1203,43 +1203,54 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
+  }
+
+  // Walk users of wide canonical IVs and apply Fn to all compares of the form
+  // (ICMP_ULE, WideCanonicalIV, backedge-taken-count).
```
Done, thanks!

fhahn committed:
!fixup address latest comments, thanks!

fhahn reviewed (commented):
Addressed comments, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1271,7 +1280,10 @@ void VPlanTransforms::addActiveLaneMask(
-  replaceHeaderPredicateWith(Plan, *LaneMask);
+  for (VPValue *HeaderMask : collectAllHeaderMasks(Plan)) {
+    HeaderMask->replaceAllUsesWith(LaneMask);
+    recursivelyDeleteDeadRecipes(HeaderMask);
```
```suggestion
```
Wonder if this is redundant due to subsequent VPlan dce.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1203,43 +1203,52 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
+  assert(count_if(Plan.getCanonicalIV()->users(),
+                  [](VPUser *U) { return isa<VPWidenCanonicalIVRecipe>(U); }) <=
+             1 &&
+         "Must at most one VPWideCanonicalIVRecipe");
```
```suggestion
         "Must be at most one VPWideCanonicalIVRecipe");
```

ayalz reviewed (approved):
This LGTM, thanks for accommodating!
Adding last minor nits.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-split-vpmemory

fhahn committed:
!fixup address final comments, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9383,24 +9337,7 @@ void VPWidenLoadRecipe::execute(VPTransformState &State) {
-          Builder, DataTy, State.get(getAddr(), Part, !CreateGather),
-          CreateGather, Mask, EVL, Alignment);
-    } else if (CreateGather) {
+    if (CreateGather) {
```
ok, but let the store case below be consistent with the load case here?

fhahn commented:
Just shared a PR to introduce an abstract header mask early on: https://github.com/llvm/llvm-project/pull/89603
